src/sisl/physics/state.py

```diff
@@ -614,17 +622,26 @@ def inner(self, ket=None, matrix=None, diag=True):
         matrix : array_like, optional
            whether a matrix is sandwiched between the bra and ket, defaults to the identity matrix.
            1D arrays will be treated as a diagonal matrix.
-        diag : bool, optional
-           only return the diagonal matrix :math:`\mathbf A_{ii}`.
+        projection:
+            how to perform the final projection.
+            This can be used to sum specific sub-elements, return the diagonal, or the
+            full matrix.
+
+            * ``diag`` only return the diagonal of the inner product
+            * ``matrix`` return a matrix of inner products, also the off-diagonals
+            * ``atoms`` only do inner products for same states, summed over atoms on the state
+            * ``states`` only do inner products for same states
```

Comment:
I find this a bit confusing. `matrix` would give the state-resolved quantity of shape `(ns, ns)`, where `ns = len(self)` while this mode returns a quantity of shape `(ns, no`). Shouldn't it rather be labeled `orbitals` (each state is resolved orbitally, no off-diagonals between states)?

Reply:
the problem is that the `state` code is generic, so no orbitals. For phonons, this is cartesian directions.
I don't really know what to name it? Suggestions?
- `states`
- `dof`
- ?

Reply:
Maybe `basis` would be a better term to avoid confusion with the "states" themselves?

The docstring could read:
* `atoms` only do inner products for individual states, but return them atom-resolved 
* `basis` only do inner products for individual states, but return them basis-resolved

In the most general case, could one also have offdiagonals (*between* states) resolved per atom- or basis?

Reply:
much better, thanks! I'll amend!

For the general case, I think that would result in too much information, no? `(nstates, nbasis, nbasis)` ? For 1000 orbital systems, this results in a matrix of 14 GB! I'll leave it out for now. But thanks for the consideration.